src/java.base/share/classes/java/math/BigDecimal.java

```diff
@@ -5219,61 +5215,130 @@ private static boolean needIncrement(MutableBigInteger mdivisor, int roundingMod
         return commonNeedIncrement(roundingMode, qsign, cmpFracHalf, mq.isOdd());
     }
 
+    /**
+     * {@code FIVE_TO_2_TO[n] == 5^(2^n)}
+     */
+    private static final BigInteger[] FIVE_TO_2_TO = new BigInteger[16 + 1];
+
+    static {
+        BigInteger pow = FIVE_TO_2_TO[0] = BigInteger.valueOf(5L);
+        for (int i = 1; i < FIVE_TO_2_TO.length; i++)
+            FIVE_TO_2_TO[i] = pow = pow.multiply(pow);
+    }
+
+    /**
+     * @param n a non-negative integer
+     * @return {@code 5^(2^n)}
+     */
+    private static BigInteger fiveToTwoToThe(int n) {
+        int i = Math.min(n, FIVE_TO_2_TO.length - 1);
+        BigInteger pow = FIVE_TO_2_TO[i];
+        for (; i < n; i++)
+            pow = pow.multiply(pow);
+
+        return pow;
+    }
+
+    private static final double LOG_5_OF_2 = Math.log(2.0) / Math.log(5.0);
+
     /**
      * Remove insignificant trailing zeros from this
      * {@code BigInteger} value until the preferred scale is reached or no
      * more zeros can be removed.  If the preferred scale is less than
      * Integer.MIN_VALUE, all the trailing zeros will be removed.
+     * Assumes {@code intVal != 0}.
      *
      * @return new {@code BigDecimal} with a scale possibly reduced
      * to be closed to the preferred scale.
      * @throws ArithmeticException if scale overflows.
      */
     private static BigDecimal createAndStripZerosToMatchScale(BigInteger intVal, int scale, long preferredScale) {
+        // avoid overflow of scale - preferredScale
+        preferredScale = Math.clamp(preferredScale, Integer.MIN_VALUE - 1L, Integer.MAX_VALUE);
+        int powsOf2 = intVal.getLowestSetBit();
+        // scale - preferredScale >= remainingZeros >= max{n : (intVal % 10^n) == 0 && n <= scale - preferredScale}
+        // a multiple of 10^n must be a multiple of 2^n
+        long remainingZeros = Math.min(scale - preferredScale, powsOf2);
+        if (remainingZeros <= 0L)
+            return valueOf(intVal, scale, 0);
+
+        final int sign = intVal.signum;
+        if (sign < 0)
+            intVal = intVal.negate(); // speed up computation of shiftRight() and bitLength()
+
+        intVal = intVal.shiftRight(powsOf2); // remove powers of 2
+        // maxPowsOf5 >= floor(log5(intVal)) >= max{n : (intVal % 5^n) == 0}
```

Comment:
```suggestion
        // Let k = max{n : (intVal % 5^n) == 0}, m = max{n : 5^n <= intVal}, so m >= k.
        // Let b = intVal.bitLength(). It can be shown that
        // | b * LOG_5_OF_2 - b log5(2) | < 2^(-21) (fp viz. real arithmetic),
        // which entails m <= maxPowsOf5 <= m + 1, where maxPowsOf5 is as below.
        // Hence, maxPowsOf5 >= k and is never off by more than 1 from the theoretical m.
```

Reply:
@rgiulietti Good, but I would not put the inequality `maxPowsOf5 <= m + 1` and say that `maxPowsOf5` is never off by more than 1 from the theoretical `m`, because it is not crucial as `maxPowsOf5 >= k`, and because `m` is in function of `intVal`, while `maxPowsOf5` is in function of `2^b`, so it is not obvious that there's no power of five `pow` such that `intVal < pow < 2^b`...

Reply:
What's really crucial for _correctness_ is to ensure maxPowsOf5 >= k.

But for performance you also want maxPowsOf5 to be as small as possible. So, the fact that it turns out that maxPowsOf5 <= m + 1 guarantees that maxPowsOf5 is the best value that can be computed very efficiently. It's more a "quality of service" guarantee than anything fundamental.

Reply:
Perhaps leave m <= maxPowsOf5 <= m + 1 and maxPowsOf5 >= k and drop the note "and is never off by more than 1 from the theoretical m"

Reply:
> What's really crucial for _correctness_ is to ensure maxPowsOf5 >= k.

Yes, I meant that the only way we know to ensure that condition is to ensure `m <= maxPowsOf5`...

> Perhaps leave m <= maxPowsOf5 <= m + 1 and maxPowsOf5 >= k and drop the note "and is never off by more than 1 from the theoretical m"

I would put `maxPowsOf5 <= m + 2` instead, because `maxPowsOf5 <= m + 1` is not obvious to prove mathematically...

Reply:
It can be proven by considering

- the inequality | b * LOG_5_OF_2 - b log5(2) | < 2^(-21)
- the properties of round() (to an integer)
- log5(2) < 1/2

The last one is crucial for the upper bound to be m + 1 rather than m + 2.

Feel free to use m + 2, though.

Reply:
I understood that it follows from the fact that `log5(2^(n+1)) - log5(2^n) == [log5(2^n) + log5(2)] - log5(2^n)`, but I think it's superfluous having to explain it in the comments just to prove `maxPowsOf5 <= m + 1`, while it is more clear to understand why `maxPowsOf5 <= m + 2` is true. Anyway, thanks for the insight.

Reply:
With the notation above, the reasoning goes like so:

2^(b-1) <= v < 5^(m+1), hence b log5(2) < m + 1 + log5(2).
This implies b * LOG_5_OF_2 < m + 1 + log5(2) + 2^(-21) < m + 1 + 1/2,
which leads to round(b * LOG_5_OF_2) <= m + 1.

As said, I don't mind if you prefer m + 2.